src/commands/deploy/setPreDeployTaskForDotnet.ts

```diff
@@ -0,0 +1,160 @@
+
+    // if we found a .csproj file set the tasks and workspace settings
+    if (csprojFile) {
+        const targetFramework: string | undefined = await tryGetTargetFramework(csprojFile);
+        context.telemetry.properties.tfw = targetFramework ? targetFramework : 'N/A';
+        if (!targetFramework) {
+            // if the target framework cannot be found, don't try to set defaults
+            return;
+        }
+
+        const notConfiguredForDeploy: string = `Required assets to build and deploy are missing from "${context.workspace.name}"? Add them?`;
+        const dontShowAgainButton: MessageItem = { title: "No, and don't show again" };
+        const input: MessageItem = await ext.ui.showWarningMessage(notConfiguredForDeploy, { modal: true }, DialogResponses.yes, dontShowAgainButton);
+        if (input === dontShowAgainButton) {
+            await updateWorkspaceSetting(showPreDeployWarningSetting, false, workspaceFspath);
+        } else {
+            // resolves to "."if it is not a subfolder
+            const subfolder: string = path.dirname(path.relative(workspaceFspath, csprojFile));
+
+            // always use posix for debug config
+            const deploySubpath: string = path.posix.join(subfolder, 'bin', 'Release', targetFramework, 'publish');
+
+            await updateWorkspaceSetting(preDeployTaskSetting, 'publish', workspaceFspath);
+            await updateWorkspaceSetting(constants.configurationSettings.deploySubpath, deploySubpath, workspaceFspath);
+
+            // update the deployContext.deployFsPath with the .NET output path since getDeployFsPath is called prior to this
+            context.deployFsPath = path.join(workspaceFspath, deploySubpath);
+
+            const existingTasks: tasks.ITask[] = tasks.getTasks(context.workspace);
+
+            // do not overwrite any dotnet commands the user may have
+            let dotnetTasks: tasks.ITask[] = generateDotnetTasks(subfolder);
+            dotnetTasks = dotnetTasks.filter(t1 => !existingTasks.find(t2 => {
+                return t1.label === t2.label;
+            }));
+
+            const newTasks: tasks.ITask[] = existingTasks.concat(dotnetTasks);
+
+            tasks.updateTasks(context.workspace, newTasks);
+        }
+
+    }
+}
+
+async function tryGetCsprojFile(context: IDeployWizardContext, projectPath: string): Promise<string | undefined> {
+    let projectFiles: string[] = await checkFolderForCsproj(projectPath);
+    // it's a common pattern to have the .csproj file in a subfolder so check one level deeper
+    if (projectFiles.length === 0) {
+        const subfolders: string[] = await fse.readdir(projectPath);
+        await Promise.all(subfolders.map(async folder => {
+            const filePath: string = path.join(projectPath, folder);
+            // check its existence as this will check .vscode even if the project doesn't contain that folder
+            if (fse.existsSync(filePath) && (await fse.stat(filePath)).isDirectory()) {
```

Comment:
Just to understand, do we want to use the async methods due to UI reasons?

Reply:
Here's a random blog post I found that explains why you should always use `async` if you can:

> When JavaScript is executed, synchronous code has the potential to block further execution until it has finished what it’s doing. In English, long-running JavaScript functions can make the UI or server unresponsive until the function has returned. Obviously this can result in a terrible user-experience.
> 
> For example: if you want to load your latest tweets onto a web page, and you do this synchronously, then a visitor to your site won’t be able to do anything until those tweets are loaded. This could cause a long delay before they even get to see the content of your site!

https://rowanmanning.com/posts/javascript-for-beginners-async/